ledger/types/balance.go

```diff
@@ -30,211 +22,91 @@ func (b *Balance) UnmarshalBinary(data []byte) error {
+	if asset != "" {
+		return &Balance{
+			Asset:  asset,
+			Amount: amount,
+		}
 	}
+	return &Balance{}
 }
 
-// Copy returns a copy of a Balance
-func (b *Balance) Copy() *Balance {
-	return &Balance{
-		UserID:     b.UserID,
-		AssetID:    b.AssetID,
-		Balance:    b.Balance,
-		Borrowed:   b.Borrowed,
-		Collateral: b.Collateral,
-		Supplied:   b.Supplied,
-		Virtual:    b.Virtual,
-		Locked:     b.Locked,
-	}
+// Valid will only be true if Balance was constructed properly
+func (b *Balance) Valid() bool {
+	return b.Asset != ""
 }
 
-// Match returns true if Balance and Amount have the same AssetID
-func (b *Balance) Match(a Amount) bool {
-	return a.AssetID == b.AssetID
+func (b *Balance) String() string {
+	return fmt.Sprintf("%d %s", b.Amount, b.Asset)
 }
 
-// Add an Amount to Balance.Balance.
-func (b *Balance) Add(amount Amount) error {
-	if !b.Match(amount) {
-		return errors.New(errors.InternalError, "Balance.Add: AssetIDs did not match")
-	}
-
-	result, err := math.CheckedAddU64(b.Balance, amount.Amount)
-	if err != nil {
-		return err
-	}
-	b.Balance = result
-	return nil
+func (b *Balance) Copy() *Balance {
+	return NewBalance(b.Asset, b.Amount)
 }
 
-// Sub an Amount from Balance.Balance.
-func (b *Balance) Sub(amount Amount) error {
-	if !b.Match(amount) {
-		return errors.New(errors.InternalError, "Balance.Add: AssetIDs did not match")
-	}
-
-	result, err := math.CheckedSubU64(b.Balance, amount.Amount)
-	if err != nil {
-		return err
-	}
-	b.Balance = result
-	return nil
+func (b *Balance) Amt() int64 {
+	return int64(b.Amount)
 }
 
-// Lock adds an Amount to Balance.Locked. Returns an error if the result Balance.Locked
-// is greater than the whole Balance.Balance
-func (b *Balance) Lock(amount Amount) error {
-	balance, err := math.CheckedSubU64(b.Balance, amount.Amount)
-	if err != nil {
-		return err
-	}
-	locked, err := math.CheckedAddU64(b.Locked, amount.Amount)
-	if err != nil {
-		return err
-	}
-	b.Balance = balance
-	b.Locked = locked
-	return nil
+func (b *Balance) AmtFloat() float64 {
+	return float64(b.Amount)
 }
 
-// Unlock subs an Amount from Balance.Locked until reach zero.
-func (b *Balance) Unlock(amount Amount) error {
-	locked, subbed := math.CheckedSubU64ToZero(b.Locked, amount.Amount)
-	balance, err := math.CheckedAddU64(b.Balance, subbed)
-	if err != nil {
-		return err
-	}
-	b.Balance = balance
-	b.Locked = locked
-	return nil
+func (b *Balance) IsZero() bool {
+	return b.Amount == 0
 }
 
-// Supply an Amount from Balance.Balance to Balance.Supplied.
-// Returns an error if not sufficient Balance.Balance.
-func (b *Balance) Supply(amount Amount) error {
-	balance, err := math.CheckedSubU64(b.Balance, amount.Amount)
-	if err != nil {
-		return err
-	}
-	supplied, err := math.CheckedAddU64(b.Supplied, amount.Amount)
-	if err != nil {
-		return err
-	}
-	b.Balance = balance
-	b.Supplied = supplied
-	return nil
+func (b *Balance) IsPositive() bool {
+	return b.Amount > 0
 }
```

Comment:
The `Amount` field is uint64, this will always be true. You cannot create a negative balance. I don't think we want to allow negative balances in the system so this is not necessary.

Reply:
We allow negative balances on `Balances` struct for borrowed values. But it's true, we don't need this function here